go-controller/pkg/ovn/logical_switch_cache.go

```diff
@@ -0,0 +1,217 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	ipam "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/ipallocator"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/ipallocator/allocator"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"k8s.io/klog"
+)
+
+// logicalSwitchInfo contains information corresponding to the node. It holds the
+// subnet allocations (v4 and v6) as well as the IPAM allocator instances for each
+// subnet managed for this node
+type logicalSwitchInfo struct {
+	hostSubnets  []*net.IPNet
+	ipams        []ipam.Interface
+	noHostSubnet bool
+}
+
+type ipamFactoryFunc func(*net.IPNet) (ipam.Interface, error)
+
+type logicalSwitchManager struct {
+	cache map[string]logicalSwitchInfo
+	// A RW mutex for logicalSwitchManager which holds logicalSwitch information
+	sync.RWMutex
+	ipamFunc ipamFactoryFunc
+}
+
+func NewIPAMAllocator(cidr *net.IPNet) (ipam.Interface, error) {
+	subnetRange, err := ipam.NewAllocatorCIDRRange(cidr, func(max int, rangeSpec string) (allocator.Interface, error) {
+		return allocator.NewContiguousAllocationMap(max, rangeSpec), nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := reserveIPs(cidr, subnetRange); err != nil {
+		klog.Warningf("Failed reserving IPs for subnet %s, err: %v", cidr, err)
+		return nil, err
+	}
+	return subnetRange, nil
+}
+
+func reserveIPs(subnet *net.IPNet, ipam ipam.Interface) error {
+	gwIfAddr := util.GetNodeGatewayIfAddr(subnet)
+	err := ipam.Allocate(gwIfAddr.IP)
+	if err != nil {
+		klog.Warningf("Unable to allocate subnet's gateway IP: %s", gwIfAddr.IP)
+		return err
+	}
+	mgmtIfAddr := util.GetNodeManagementIfAddr(subnet)
+	err = ipam.Allocate(mgmtIfAddr.IP)
+	if err != nil {
+		klog.Warningf("Unable to allocate subnet's management IP: %s", mgmtIfAddr.IP)
+		return err
+	}
+	if config.HybridOverlay.Enabled {
+		hybridOverlayIfAddr := util.GetNodeHybridOverlayIfAddr(subnet)
+
+		err = ipam.Allocate(hybridOverlayIfAddr.IP)
+		if err != nil {
+			klog.Warningf("Unable to allocate subnet's hybrid overlay interface IP: %s", hybridOverlayIfAddr.IP)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func newLogicalSwitchManager() *logicalSwitchManager {
+	return &logicalSwitchManager{
+		cache:    make(map[string]logicalSwitchInfo),
+		RWMutex:  sync.RWMutex{},
+		ipamFunc: NewIPAMAllocator,
+	}
+}
+
+func (manager *logicalSwitchManager) Add(nodeName string, hostSubnets []*net.IPNet) error {
```

Comment:
Naming suggestion: `AddNode()`

Reply:
question: How does this work for dual-stack? Each address family can have multiple hostSubnets.

Reply:
No, the only place we support more than just "1 IPv4 and 1 IPv6" for _anything_ is that you can have multiple cluster CIDRs of each type. But any given node will only get assigned a single subnet per IP family.